rplugin/python3/deoplete/sources/tag.py

```diff
@@ -24,31 +24,46 @@
 #=============================================================================
 
 import re
+from os.path import getmtime
+from collections import namedtuple
 from .base import Base
 
+TagsCacheItem = namedtuple("TagsCacheItem", "mtime tags")
+
 class Source(Base):
     def __init__(self):
         Base.__init__(self)
 
         self.mark = '[T]'
 
+        self.cache = {}
+
     def gather_candidates(self, vim, context):
         candidates = []
-
         for tags_file in vim.eval(
                 'map(tagfiles(), "fnamemodify(v:val, \\":p\\")")'):
-            with open(tags_file, 'r', errors='replace') as f:
-                candidates += parse_tags(f)
-        return [{ 'word': x } for x in list(set(candidates))]
+            mtime = getmtime(tags_file)
+            if tags_file not in self.cache or \
+                    self.cache[tags_file].mtime != mtime:
+                with open(tags_file, 'r', errors='replace') as f:
+                    candidates += parse_tags(f)
+                self.cache[tags_file] = TagsCacheItem(mtime, candidates)
+            else:
+                candidates += self.cache[tags_file].tags
+        return [{ 'word': x } for x in candidates]
 
 def debug(vim, msg):
     vim.command('echomsg string("' + str(msg) + '")')
 
 def parse_tags(f):
-    p = re.compile('^[a-zA-Z_]\w*(?=\t)')
+    p = re.compile('[a-zA-Z_]\w*(?=\t)')
```

Comment:
Why? This variant is going to catch more things as tags. In some cases it will *definitely* catch non-tags or partial tags. And this change also

1. Has absolutely nothing to do with adding caching (commit 1, PR purpose).
2. Has absolutely nothing to do with changing method of avoiding candidates (commit 2 in which it is present).